lib/iso8601/time_interval.rb

```diff
@@ -0,0 +1,367 @@
+module ISO8601
+  ##
+  # A Time Interval representation.
+  # See https://en.wikipedia.org/wiki/ISO_8601#Time_intervals
+  #
+  # @example
+  #     ti = ISO8601::TimeInterval.new('P1MT2H/2014-05-28T19:53Z')
+  #     ti.to_f # => 2635200.0
+  #     ti2 = ISO8601::TimeInterval.new('2014-05-28T19:53Z/2014-05-28T20:53Z')
+  #     ti2.to_f # => 3600.0
+  #
+  # @example
+  #     start_time = ISO8601::DateTime.new('2014-05-28T19:53Z')
+  #     end_time = ISO8601::DateTime.new('2014-05-30T19:53Z')
+  #     ti = ISO8601::TimeInterval.from_datetimes(start_time, end_time)
+  #     ti.to_f # => 172800.0 (Seconds)
+  #
+  # @example
+  #     duration = ISO8601::Duration.new('P1MT2H')
+  #     end_time = ISO8601::DateTime.new('2014-05-30T19:53Z')
+  #     ti = ISO8601::TimeInterval.from_duration(duration, end_time: end_time)
+  #     ti.to_f # => 2635200.0 (Seconds)
+  #
+  # @example
+  #     start_time = ISO8601::DateTime.new('2014-05-30T19:53Z')
+  #     duration = ISO8601::Duration.new('P1MT2H', base)
+  #     ti = ISO8601::TimeInterval.from_duration(duration, start_time: start_time)
+  #     ti.to_f # => 2635200.0 (Seconds)
+  #
+  class TimeInterval
+    include Comparable
+
+    # Define the type of a time
+    TYPE_DATETIME = :datetime
+    TYPE_DURATION = :duration
+
+    ##
+    # Initialize a TimeInterval based on a two ISO8601::DateTime instances.
+    #
+    # @param [ISO8601::DateTime] start_time An ISO8601::DateTime that represents
+    #     start time of the interval.
+    # @param [ISO8601::DateTime] start_time An ISO8601::DateTime that represents
+    #     end time of the interval.
+    #
+    # @raise [ArgumentError] If any param is not an instance of ISO8601::DateTime
+    #
+    def self.from_datetimes(start_time, end_time)
+      valid_date_time?(start_time) && valid_date_time?(end_time)
+      new("#{start_time}/#{end_time}")
+    end
+
+    ##
+    # Initialize a TimeInterval based on a ISO8601::Duration. To set a correct
+    # time interval, we need a start or end point. This point is sent in second
+    # parameter.
+    #
+    # @param [ISO8601::Duration] duration An ISO8601::Duration that represent the
+    #     size of the interval.
+    # @param [Hash] time A hash to set the start or end point of the time interval.
+    #     This hash must contain an unique key (start_time or end_time) and a
+    #     ISO8601::DateTime as value. For example:
+    #       { start_time: iso_8601_datetime }
+    #       { end_time: iso_8601_datetime }
+    #
+    # @raise [ArgumentError] if duration is not an instance of ISO8601::Duration
+    # @raise [ArgumentError] if keys of time hash are not valid, or the value is not
+    #     an instance of ISO8601::DateTime
+    #
+    def self.from_duration(duration, time)
```

Comment:
I prefer the narrative to be more like: check params, set internals, happy path.

```ruby
def self.from_duration(duration, time)
  valid_duration?(duration)
  fail(ArgumentError, 'Time param ....') unless time.is_a?(Hash) && time.keys.size == 1
  start_time = time[:start_time]
  end_time = time[:end_time]
  fail(ArgumentError, 'You must...') if start_time.nil? && end_time.nil?
  
  pattern = if start_time
    valid_date_time?(start_time)
    "#{start_time}/#{duration}"
  else
    valid_date_time?(end_time)
    "#{duration}/#{end_time}"
  end

  new(pattern)
end
```